source/common/upstream/upstream_impl.cc

```diff
@@ -987,7 +988,18 @@ void MainPrioritySetImpl::updateCrossPriorityHostMap(const HostVector& hosts_add
   }
 
   for (const auto& host : hosts_removed) {
-    mutable_cross_priority_host_map_->erase(addressToString(host->address()));
+    auto existing_host = mutable_cross_priority_host_map_->find(addressToString(host->address()));
+
+    if (existing_host != mutable_cross_priority_host_map_->end()) {
+      // an earlier priority added this host, which indicates that the host moved from
+      // a higher value priority to a lower value priority. We should not remove it from the
+      // cross priority map.
+      if (existing_host->second->priority() < priority) {
+        continue;
+      } else {
+        mutable_cross_priority_host_map_->erase(addressToString(host->address()));
+      }
```

Comment:
Please convert to `if (existing_host->second->priority() >= priority) ...` and remove the `continue` part.